mllib/src/test/scala/org/apache/spark/ml/feature/QuantileDiscretizerSuite.scala

```diff
@@ -82,14 +82,23 @@ class QuantileDiscretizerSuite
       .setInputCol("input")
       .setOutputCol("result")
       .setNumBuckets(numBuckets)
-
-    // Reserve extra one bucket for NaN
-    val expectedNumBuckets = discretizer.fit(df).getSplits.length - 1
-    val result = discretizer.fit(df).transform(df)
-    val observedNumBuckets = result.select("result").distinct.count
-    assert(observedNumBuckets == expectedNumBuckets,
-      s"Observed number of buckets are not correct." +
-        s" Expected $expectedNumBuckets but found $observedNumBuckets")
+    val splits = Array(Double.NegativeInfinity, 1.0, Double.PositiveInfinity)
+
+    for (handleNaN <- Array("keep", "skip")) {
+      discretizer.setHandleNaN(handleNaN)
+      val expectedNumBuckets = {
```

Comment:
This is kind of complex logic just to get the expected number of splits.  I'd recommend just putting the expected bucket *values* in the original DataFrame:
```
val df = sc.parallelize(Seq(
  (1.0, /*expected value for option "keep"*/, /*expected value for option "skip"*/),
  ...
)).toDF("input", "expectedKeep", "expectedSkip")
```
Then you can compare with the actual values.  That'll be an easier test to understand IMO.